ert_storage/testing/testclient.py

```diff
@@ -0,0 +1,276 @@
+        check: bool = True,
+    ) -> dict:
+        doc = self.gql_client.execute(request_string, variable_values=variable_values)
+        if check and "errors" in doc:
+            raise ClientError(f"GraphQL query returned an error:\n{pformat(doc)}")
+
+        return doc
+
+
+@contextmanager
+def testclient_factory() -> Generator[_TestClient, None, None]:
+    if "ERT_STORAGE_DATABASE_URL" not in os.environ:
+        os.environ["ERT_STORAGE_DATABASE_URL"] = "sqlite:///:memory:"
+        print("Using in-memory SQLite database for tests")
+
+    from ert_storage.app import app
+    from ert_storage.graphql import schema
+
+    session, transaction, connection = _begin_transaction()
+    schema.override_session = session
+
+    yield _TestClient(app, session=session, gql_schema=schema)
+
+    schema.override_session = None
+    _end_transaction(transaction, connection)
+
+
+_TransactionInfo = Tuple[sessionmaker, Transaction, Any]
+
+
+def _override_get_db(session: sessionmaker) -> None:
+    from ert_storage.app import app
+    from ert_storage.database import (
+        get_db,
+        IS_POSTGRES,
+    )
+
+    async def override_get_db() -> AsyncGenerator[Session, None]:
+        db = session()
+
+        # Make PostgreSQL return float8 columns with highest precision. If we don't
+        # do this, we may lose up to 3 of the least significant digits.
+        if IS_POSTGRES:
+            db.execute("SET extra_float_digits=3")
+        try:
+            yield db
+            db.commit()
+            db.close()
+        except:
+            db.rollback()
+            db.close()
+            raise
+
+    app.dependency_overrides[get_db] = override_get_db
+
+
+def _begin_transaction() -> _TransactionInfo:
+    from ert_storage.database import (
+        engine,
+        IS_SQLITE,
+        HAS_AZURE_BLOB_STORAGE,
+    )
+    from ert_storage.database_schema import Base
+
+    if IS_SQLITE:
+        Base.metadata.create_all(bind=engine)
+    if HAS_AZURE_BLOB_STORAGE:
+        import asyncio
+        from ert_storage.database import create_container_if_not_exist
+
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(create_container_if_not_exist())
+
+    connection = engine.connect()
+    transaction = connection.begin()
+    session = sessionmaker(autocommit=False, autoflush=False, bind=connection)
+
+    _override_get_db(session)
+
+    return session, transaction, connection
+
+
+def _end_transaction(transaction: Transaction, connection: Any) -> None:
+    # teardown: rollback database to before the test.
+    # For debugging change rollback to commit.
+    transaction.rollback()
+    connection.close()
+
+
+def _check(check_status_code: Optional[int], response: requests.Response) -> None:
+    if check_status_code is not None and response.status_code != check_status_code:
+        try:
+            doc = response.json()
+        except:
+            doc = response.content
+        raise ClientError(
```

Comment:
Was wondering here whether `raise` shouldn't be in the except branch?

Reply:
I might not get it correct, but isn't this check ( the `if` ) checking for error, and the `doc` either gets the error from the json or from the exception when trying to retrieve the json? And then it ends with raising an exception writing out the error?

Reply:
Yep. I want to display the response with prettyprinter's `pformat`. So we test to see if it's a JSON, and then we fallback to just displaying the `bytes` of the response.